### PR TITLE
Document cost-based join reordering

### PR DESCRIPTION
**Pull request by rschlussel-zz**

### CODE REVIEW

File: presto-docs/src/main/sphinx/optimizer/reorder-joins.rst [hunk 1]
```diff
@@ -63,8 +65,19 @@ For the following query:
...
 The join order will change from ``part, orders, lineitem, supplier, nation`` to
 ``part, lineitem, orders, supplier, nation``.
+
+
+COST_BASED
+==========
+
+When the join reordering strategy is set to ``COST_BASED``, Presto will use statistics
+provided by the connectors to estimate the costs for different join orders (currently stats
+are only available in the hive connector). It will then choose the join order with the lowest
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/optimizer/reorder-joins.rst:80]: s/hive/Hive

**Review by akshatnair — APPROVED** [review]
LGTM mod a small change

File: presto-docs/src/main/sphinx/optimizer/reorder-joins.rst [hunk 2]
```diff
@@ -63,8 +65,19 @@ For the following query:
...
+ WHERE p.partkey = l.partkey AND l.orderkey = o.orderkey AND l.suppkey = s.suppkey AND s.nationkey = n.nationkey;
 
 The join order will change from ``part, orders, lineitem, supplier, nation`` to
 ``part, lineitem, orders, supplier, nation``.
+
+
+COST_BASED
+==========
+
+When the join reordering strategy is set to ``COST_BASED``, Presto will use statistics
```
Comment (kbajda) [inline presto-docs/src/main/sphinx/optimizer/reorder-joins.rst:78]: Let's link to a section of docs describing Stats in Presto. How to gather them, display etc.

File: presto-docs/src/main/sphinx/optimizer/reorder-joins.rst [hunk 3]
```diff
@@ -63,8 +65,19 @@ For the following query:
...
+provided by the connectors to estimate the costs for different join orders (currently stats
+are only available in the hive connector). It will then choose the join order with the lowest
+computed cost. If statistics are not available or if for any other reason a cost could not be
+computed, the ``ELIMINATE_CROSS_JOINS`` strategy is used instead. Enumerating all possible join
+orders for many tables is computationally intensive, so joins are reordered in groups of ten at a time.
```
Comment (kbajda) [inline presto-docs/src/main/sphinx/optimizer/reorder-joins.rst:83]: Could you explain what happens if a query has more than 10 joins, say 12? Would 10 first joins be reordered between them and then the remaining 2 will be reordered separately? and the clusters of 10 and 2 will be merged without considering any cost between the two groups?

**Review by kbajda — CHANGES_REQUESTED** [review]
Please clarify what about the replicated joins? It used to be a separate property independent of "cross join elimination".  Is "cost-based" doing that automatically now? What if someone selects "eliminate-cross-joins" instead? any chance for them to get replicated joins via setting another property?

File: presto-docs/src/main/sphinx/optimizer/reorder-joins.rst (hunk 3 above)
Comment (rschlussel-zz) [reply presto-docs/src/main/sphinx/optimizer/reorder-joins.rst:83]: yes.

**Comment by rschlussel-zz** [comment]
Repartitioned vs. replicated vs. automatic is still a separate property that is enforced regardless of the join reordering strategy.  It has its own docs that haven't changed from last time.

**Comment by kbajda** [comment]
> is still a separate property
If so, please make sure we mention that and link to that doc section somehow and explain that for full cost-based benefit, one needs to enable both things.

File: presto-docs/src/main/sphinx/optimizer/reorder-joins.rst (hunk 2 above)
Comment (rschlussel-zz) [reply presto-docs/src/main/sphinx/optimizer/reorder-joins.rst:78]: I'll have to do that later when I rebase this on top of the stats and join enumeration branches.

**Comment by rschlussel-zz** [comment]
merged to  the sprint branch.

**Pull request closed without merging**